### ISSUE TITLE
Explore UI w/ Prometheus Hangs

### ISSUE THREAD

**Issue by timlharrison** [grafana/grafana#17994]
**What happened**:
Using Grafana (Windows) 6.2.5 , Prometheus (Windows) 2.10.0 on the same server and metrics from CF exporters I'm experiencing very poor performance with the Explore UI and constant crashes.  I'm trying to use it the same way I use the Prometheus UI - typing queries and changing queries as I type them, etc.  Chrome dev tools show it trying to download hundreds of MB of data from "https://<grafana>:8080/api/datasources/proxy/1/api/v1/series?match[]={__name__=%22cf_application_info%22}"

**What you expected to happen**:
I expect the UI to wait until I click "Run Query" but it seems to be trying to pre-load massive amounts of information as I type it causing the browser to hang.

**How to reproduce it (as minimally and precisely as possible)**:
Not sure if it can be reproduced easily... basically just trying to use the Explore UI at all in an environment like this with lots of CF metrics.  In this case all I did was type "cf_application_info" and then started trying to use some filters within curly braces.

**Anything else we need to know?**:
It seems to me that the issue arises when its trying to load label values or something along those lines while typing.  It doesn't happen on clicking "Run Query".  Also, when this occurs, I see the memory usage on the Prometheus server double or more.

**Environment**:
- Grafana version: 6.2.5
- Data source type & version: Prometheus 2.10.0
- OS Grafana is installed on: Windows
- User OS & Browser: Win10/Chrome 75.0.3770.100
- Grafana plugins:
- Others:

**Comment by FUSAKLA** [comment]
Hi, I believe the fix was merget to the master yesterday. The issue looks fairly similar. 
See https://github.com/grafana/grafana/issues/16844#event-2466620591

**Comment by dprokop** [comment]
@timlharrison can you try out latest build and see whether or not the problem still occurs (https://grafana.com/grafana/download/6.3.0-765f5c4cpre)?

**Comment by timlharrison** [comment]
I tested in a Windows 2019 Docker container and the good news is it didn't crash - the bad news is it was still kind of slow, but that could just be because it was running in docker on my workstation.  I'll update after 6.3 goes GA and we get it running in Prod.  Any ETA on that?

**Comment by FUSAKLA** [comment]
Yeah, well before it queried all series you have from whole retention period. Now it should query just the specified time range as set in Grafana. If you have high cardinality data and lots of time series than it can be still slow. Also it depends on the size of the time range.

**Comment by timlharrison** [comment]
Is there any way to disable the hints? Maybe that's a good solution for folks like me.

**Comment by FUSAKLA** [comment]
Yeah that could be handy. IIRC it was suggested in some related issue but not sure how that ended up.

**Comment by ahurtaud** [comment]
It would be great to have an update on this.
With 6.3 as it was added to the query editor panel, it just hang the browser when we edit dashboards now :/.

PS: We use thanos datasource so we have a massive amount of metrics names...

**Comment by pablokbs** [comment]
I'm having the same issue, with the new Edit UI it takes about 30 seconds to finish "Loading metrics"

![image](https://user-images.githubusercontent.com/4028336/65251540-e2384480-dacd-11e9-80e1-a87509cdf009.png)

I had to roll back to Grafana 6.2.5

**Comment by ahayworth** [comment]
We can confirm this problem at GitHub. We have certain series with extremely high cardinality (and while we have plans to address the cardinality problems at the source, we cannot do it immediately nor can we guarantee it will never happen again).

For us, the most problematic aspect is that the Explore UI ends up hanging the browser as people are typing due to large API requests that can take multiple seconds to complete (in particularly bad cases, almost 60s).

For example, if you type something like `up{}` into the explore UI on our installation, the frontend requests:

```
/api/datasources/proxy/1/api/v1/series?match[]={__name__=%22up%22}&start=1573817717&end=1573821317
```

Because we actually query Thanos rather than Prometheus (which proxies requests to many downstream datastores), this fans out to many queries. The UI autocompletion request above results in more than 20k series:

```
~/projects/grafana master
❯ jq '.data | length' /Users/ahayworth/Desktop/series.json
21641
```

That is slow-ish, but usable. A far more problematic example is typing `glb_response_bucket{}` (which is a very high cardinality metric internally). The Grafana UI  requests:

```
/api/v1/series?match[]={__name__=%22glb_response_bucket%22}&start=1573817717&end=1573821317
```

and the response is huge:

```
~/projects/grafana master
❯ jq '.data | length' /Users/ahayworth/Desktop/series-2.json
461316
```

Not only does that take a long time to generate results on Thanos; it takes a long time to transfer the resulting data (it's about 200MB) and then the UI struggles to handle it.

I'm not sure exactly what Grafana can do to mitigate performance problems in this case, but I trust that there are smarter people than me here who may have ideas. 😄 😄 

One thing that might be reasonable would be to wait longer before issuing the queries for autocompletion, or to offer the ability to turn it off.

**Comment by davkal** [comment]
Fixed via #20936
